spynnaker/pyNN/models/neuron/neuron_data.py

```diff
@@ -0,0 +1,371 @@
+# Copyright (c) 2022 The University of Manchester
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import numpy
+from spynnaker.pyNN.models.common.param_generator_data import (
+    is_param_generatable)
+from spinn_front_end_common.utilities.constants import BYTES_PER_WORD
+from spinn_front_end_common.utilities.helpful_functions import (
+    locate_memory_region_for_placement)
+from spynnaker.pyNN.utilities.struct import StructRepeat
+from spinn_utilities.helpful_functions import is_singleton
+from spynnaker.pyNN.data import SpynnakerDataView
+
+
+def _all_one_val_gen(rd):
+    """ Determine if all the values of a dictionary are the same,
+        and can be generated.  Note that a random distribution is considered
+        the same if the same distribution is used for all neurons.
+
+    :rtype: bool
+    """
+    for key in rd.keys():
+        if is_singleton(rd[key]):
+            if not is_param_generatable(rd[key]):
+                return False
+        else:
+            if not rd[key].range_based():
+                return False
+            for i, (_start, _stop, val) in enumerate(rd[key].iter_ranges()):
+                if i > 0:
+                    return False
+                if not is_param_generatable(val):
+                    return False
+    return True
+
+
+def _all_gen(rd):
+    """ Determine if all the values of a ranged dictionary can be generated.
+
+    :rtype: bool
+    """
+    for key in rd.keys():
+        if is_singleton(rd[key]):
+            if not is_param_generatable(rd[key]):
+                return False
+        else:
+            if not rd[key].range_based():
+                return False
+            for _start, _stop, val in rd[key].iter_ranges():
+                if not is_param_generatable(val):
+                    return False
+    return True
+
+
+class NeuronData(object):
```

Comment:
The term Data is already over used most importantly with information coming off the machine.

Would it not be better to use "Specs" here?

Both in the class name and in the methods.
So that what is here "write_data" (also used in Population ect) becomes write_specs

Reply:
Its all data... the things coming off the machine is recorded data... I do see the point, but I think it could do with a more global review than just here, since data is used in lots of places I think!